File: player/chromeos/src/player/remoteCommandHandler.ts
/**
 * Translates CMS RemoteCommands into VideoPlayer operations.
 *
 * Receives a RemoteCommandPayload, dispatches to the matching VideoPlayer
 * method, then acks the command via CmsControlClient. Unknown / invalid
 * commands are NAK'd with a descriptive error message.
 */
import type { VideoPlayer } from './videoPlayer';
import type {
  CmsControlClient,
  RemoteCommandPayload,
} from '../realtime/cmsControlClient';

export class RemoteCommandHandler {
  constructor(
    private readonly player: VideoPlayer,
    private readonly cmsClient: CmsControlClient,
    private readonly onForceSync: () => void
  ) {}

  async handle(cmd: RemoteCommandPayload): Promise<void> {
    const type = (cmd.commandType || '').toLowerCase();
    const payload = this.parsePayload(cmd.payloadJson);

    try {
      switch (type) {
        case 'play':
          this.player.resume();
          break;
        case 'pause':
          this.player.pause();
          break;
        case 'skip':
        case 'next':
          this.player.skipNext();
          break;
        case 'previous':
          this.player.skipPrevious();
          break;
        case 'restart':
        case 'restartloop':
          this.player.restart();
          break;
        case 'jumpto': {
          const index = payload.index ?? payload.itemIndex;
          if (typeof index !== 'number') {
            throw new Error("JumpTo requires numeric 'index'");
          }
          this.player.jumpTo(index);
          break;
        }
        case 'setvolume': {
          if (typeof payload.volume !== 'number') {
            throw new Error("SetVolume requires numeric 'volume'");
          }
          this.player.setVolume(payload.volume);
          break;
        }
        case 'forcesync':
          this.onForceSync();
          break;
        case 'reboot':
          await this.cmsClient.ackCommand(cmd.id, true);
          // ChromeOS player reboot = page reload (triggers handshake + re-auth)
          setTimeout(() => window.location.reload(), 300);
          return;
        default:
          await this.cmsClient.ackCommand(cmd.id, false, `Unknown command: ${cmd.commandType}`);
          return;
      }
      await this.cmsClient.ackCommand(cmd.id, true);
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      console.error(`[RemoteCmd] ${cmd.commandType} failed:`, msg);
      await this.cmsClient.ackCommand(cmd.id, false, msg);
    }
  }

  private parsePayload(raw: string | null): Record<string, unknown> {
    if (!raw) return {};
    try {
      const parsed = JSON.parse(raw);
      return typeof parsed === 'object' && parsed !== null ? (parsed as Record<string, unknown>) : {};
    } catch {
      return {};
    }
  }
}

File: player/raspberrypi/remote_command_handler.py
"""
Translates CMS remote commands into mpv operations for the Raspberry Pi player.

Called by `CmsControlClient._on_command` with the RemoteCommandDto payload.
After executing the command, acknowledges via `CmsControlClient.ack_command`.

Supported commands (case-insensitive): Play, Pause, Skip/Next, Previous,
Restart/RestartLoop, JumpTo, SetVolume, ForceSync, Reboot. Unknown command
types are NAK'd with a descriptive error.
"""

import json
import logging
import os
import sys
from typing import Any, Dict, Optional

logger = logging.getLogger(__name__)


class RemoteCommandHandler:
    def __init__(self, mpv_player, cms_client, on_force_sync=None):
        """
        Args:
            mpv_player: MPVDualPlayer instance (must expose player_a with mpv props)
            cms_client: CmsControlClient — used to call ack_command
            on_force_sync: optional callback invoked on ForceSync / playlist_updated
                           to trigger a re-handshake.
        """
        self.mpv_player = mpv_player
        self.cms_client = cms_client
        self.on_force_sync = on_force_sync

    def handle(self, cmd: Dict[str, Any]) -> None:
        command_id = cmd.get("id") or cmd.get("Id")
        command_type = (cmd.get("commandType") or cmd.get("CommandType") or "").lower()
        payload_json = cmd.get("payloadJson") or cmd.get("PayloadJson")
        payload = self._parse_payload(payload_json)

        try:
            if command_type == "play":
                self._set_pause(False)
            elif command_type == "pause":
                self._set_pause(True)
            elif command_type in ("skip", "next"):
                self._mpv_command(["playlist-next", "weak"])
            elif command_type == "previous":
                self._mpv_command(["playlist-prev", "weak"])
            elif command_type in ("restart", "restartloop"):
                self._mpv_set("playlist-pos", 0)
                self._set_pause(False)
            elif command_type == "jumpto":
                index = payload.get("index", payload.get("itemIndex"))
                if index is None:
                    raise ValueError("JumpTo requires 'index'")
                self._mpv_set("playlist-pos", int(index))
                self._set_pause(False)
            elif command_type == "setvolume":
                vol = payload.get("volume")
                if vol is None:
                    raise ValueError("SetVolume requires 'volume' (0-100)")
                # mpv 'volume' property is 0-100
                self._mpv_set("volume", max(0, min(100, float(vol) * 100 if float(vol) <= 1 else float(vol))))
            elif command_type == "forcesync":
                if self.on_force_sync:
                    self.on_force_sync()
            elif command_type == "reboot":
                self._ack(command_id, True)
                logger.warning("Reboot command — exiting process")
                # Systemd / watchdog will restart us; use _exit so no cleanup blocks.
                os._exit(0)
                return
            else:
                self._ack(command_id, False, f"Unknown command: {command_type}")
                return

            self._ack(command_id, True)
        except Exception as e:
            logger.exception("Command %s failed", command_type)
            self._ack(command_id, False, str(e))

    # ── Helpers ──

    def _set_pause(self, paused: bool) -> None:
        player = getattr(self.mpv_player, "active_player", None) or getattr(self.mpv_player, "player_a", None)
        if player is None:
            raise RuntimeError("mpv player not available")
        player.pause = paused

    def _mpv_command(self, args) -> None:
        player = getattr(self.mpv_player, "active_player", None) or getattr(self.mpv_player, "player_a", None)
        if player is None:
            raise RuntimeError("mpv player not available")
        player.command(*args)

    def _mpv_set(self, prop: str, value) -> None:
        player = getattr(self.mpv_player, "active_player", None) or getattr(self.mpv_player, "player_a", None)
        if player is None:
            raise RuntimeError("mpv player not available")
        setattr(player, prop, value)

    def _parse_payload(self, raw: Optional[str]) -> Dict[str, Any]:
        if not raw:
            return {}
        try:
            return json.loads(raw)
        except Exception:
            return {}

    def _ack(self, command_id: Optional[str], success: bool, error: Optional[str] = None) -> None:
        if not command_id:
            logger.warning("ack skipped — no command id")
            return
        self.cms_client.ack_command(str(command_id), success, error)
